analyses/tmb-compare-tcga/TMB_d3b_code/code/02_cumulative_freq_TMBplot.py

```diff
@@ -0,0 +1,121 @@
+##  This script takes in TMB scores outfile and calculates cumulative distribution
+#   function plot.
+# INputs  include the tmb  scores text file, outfile name and min number of samples to be plotted
+#                   (no need to use any extensions for out file, just the name)
+#   Somtimes MAF can have very few samples under each histology and those histologies can be filtered
+#       out of the plot using the min samples feature
+
+# The TMB  text  should have the following column headers -
+#   Samplename experimental_strategy disease count bedlength  TMB
+# Use results from 01_calculate_tmb_targetflexible.py script
+
+####### Author: Teja Koganti ##################
+# python3 02_cumulative_freq_TMBplot.py  \
+#   -i ../output/pbta-snv-consensus.tmb.txt \
+#   -o ../output/pbta-snv-consensus \
+#   -s 10
+
+
+import subprocess
+import argparse
+import sys
+
+# This checks the packages to be installed if not already
+#   installed by user
+def install_package(package, package_list):
+    if not package in package_list:
+        print("Installing package " + package)
+        pip._internal.main(["install", package])
+
+
+###################################################################
+############# Checking if all packages are installed ##############
+
+reqs = subprocess.check_output([sys.executable, "-m", "pip", "freeze"])
+installed_packages = [r.decode().split("==")[0] for r in reqs.split()]
+
+needed_packages = [
+    "pandas",
+    "numpy",
+    "matplotlib",
+]
+
+for package in needed_packages:
+    install_package(package, installed_packages)
+
+##################################################################
+
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import pip
+
+parser = argparse.ArgumentParser()
+parser.add_argument("-t", "--tmb_scores", required=True, help="file with TMB scores")
+parser.add_argument(
+    "-o", "--outfilename", required=True, help="Name of the out plot, no extension"
+)
+parser.add_argument(
+    "-s",
+    "--minsamplestoplot",
+    required=True,
+    help="Minimum samples from each histology/disease to plot",
+)
+args = parser.parse_args()
+
+######### Preparing inputs files and disease list ###########
+# Loading TMB scres text file into a  dataframe
+tmbfile = pd.read_csv(args.tmb_scores, sep="\t")
+
+# Choosing disease list to plot
+#tmbfile["new_TMB"] = tmbfile["TMB"]+1
```

Comment:
Why is this part commented out?

Reply:
I am implementing these lines instead - 

`value_to_add = np.min(tmbfile[tmbfile["TMB"] != 0]["TMB"])*0.8`
`tmbfile["new_TMB"] = tmbfile["TMB"].replace({0: value_to_add})`
`tmbfile = tmbfile.sort_values(by=['new_TMB'])`


When I looked at the values after adding one to all - there was not a lot of difference amongst the lower values. They did not look different on the plot. So instead of adding 1 to all values, 
- I take 80% of the smallest value (that is non-zero) 
- name it `value_to_add`
- Replace all zero values with this value - so taking log will not longer be a problem

Reply:
This seems like a bit of an arbitrary solution; I think a more principled position might be to add 1 mutation to each sample, then recalculate TMB and log(TMB)... You have all of the data to do this in the file, I think, so you could calulate something like: 

```
tmbfile["new_TMB"] = ( tmbfile["count"] + 1 * 1000000) / tmbfile["bedlength"]
```

(If we want to be very principled, we would probably add something small to the denominator as well, but that won't matter much as long as the denominator is large!)